src/catacharset.h

```diff
@@ -176,4 +176,74 @@ class utf8_wrapper
         void init_utf8_wrapper();
 };
 
+/* A range that iterates through Unicode code points in a UTF-8 encoded string
+ * without incurring dynamic memory allocation.
+ *
+ * Example:
+ *   for( char32_t c : utf8_view( "..." ) ) {
+ *       do_something_with( c );
+ *   }
+ */
+class utf8_view
+{
+    private:
+        const char *buffer;
+        std::size_t length;
+
+        class iterator
+        {
+            public:
+                using iterator_category = std::input_iterator_tag;
+                using difference_type = std::ptrdiff_t;
+                using value_type = char32_t;
+                using pointer = value_type*;
+                using reference = value_type&;
+            private:
+                const char *ptr;
+                const char *next_ptr;
+                int remaining;
+                int next_remaining;
+                char32_t unicode;
+
+                void decode() {
+                    if( remaining > 0 ) {
+                        unicode = UTF8_getch( &next_ptr, &next_remaining );
+                    } else {
+                        next_ptr = nullptr;
+                        next_remaining = 0;
+                        unicode = 0;
+                    }
+                }
+            public:
+                explicit iterator( const char *ptr, int remaining ) : ptr( ptr ), remaining( remaining ) {
+                    next_ptr = ptr;
+                    next_remaining = remaining;
+                    decode();
+                }
+                bool operator != ( const iterator &rhs ) const noexcept {
+                    return this->ptr != rhs.ptr;
```

Comment:
Would this be the same if you compare `next_ptr` instead? If so `ptr` and `remaining` can be removed.

Reply:
I'm thinking of adding the ability to return a `std::string_view` or `std::span` or something to describe a contiguous UTF-8 bytes range in the future, so I choose to keep the pointers to the beginning and the end of the UTF-8 bytes representing the current Unicode code point.